Redis connection failure raises TypeError, bypasses retry

If connecting to redis returns a timeout, kombu raises a `TypeError` instead of the timeout error and this bypasses any retry logic.

```
> lambda: self._connect(), lambda error: self.disconnect(error)
E   TypeError: disconnect() takes 1 positional argument but 2 were given
```

This is because the code [here](https://github.com/redis/redis-py/blob/b5ebada842c715e9ee74ea638a7e6d11afcddae1/redis/connection.py#L612) in redis/connection.py calls `self.disconnect(error)`:

```python
        try:
            sock = self.retry.call_with_retry(
                lambda: self._connect(), lambda error: self.disconnect(error)
            )
        except socket.timeout:
```

However in the local subclass [here](https://github.com/celery/kombu/blob/83e21c838252accb446f0f54e0d852d81cbd7f9b/kombu/transport/redis.py#L1187-L1190) in kombu, `disconnect()` does not accept any arguments:

```python
            class Connection(connection_cls):
                def disconnect(self):
                    super().disconnect()
                    channel._on_connection_disconnect(self)
            connection_cls = Connection
```

--- a/kombu/transport/redis.py
+++ b/kombu/transport/redis.py
@@ -1185,8 +1185,8 @@ class Channel(virtual.Channel):
 
         if asynchronous:
             class Connection(connection_cls):
-                def disconnect(self):
-                    super().disconnect()
+                def disconnect(self, *args):
+                    super().disconnect(*args)
                     channel._on_connection_disconnect(self)
             connection_cls = Connection
 

--- a/t/integration/test_redis.py
+++ b/t/integration/test_redis.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
 import os
+import socket
 from time import sleep
 
 import pytest
@@ -134,3 +135,18 @@ class test_RedisPriority(BasePriority):
 @pytest.mark.flaky(reruns=5, reruns_delay=2)
 class test_RedisMessage(BaseMessage):
     pass
+
+
+@pytest.mark.env('redis')
+def test_RedisConnectTimeout(monkeypatch):
+    # simulate a connection timeout for a new connection
+    def connect_timeout(self):
+        raise socket.timeout
+    monkeypatch.setattr(
+        redis.connection.Connection, "_connect", connect_timeout)
+
+    # ensure the timeout raises a TimeoutError
+    with pytest.raises(redis.exceptions.TimeoutError):
+        # note the host/port here is irrelevant because
+        # connect will raise a socket.timeout
+        kombu.Connection('redis://localhost:12345').connect()

